gson/src/main/java/com/google/gson/JsogPolicy.java

```diff
@@ -0,0 +1,183 @@
+   * @return A prefix that should be added to the instance's JSOG id
+   * @see #withTypePrefix()
+   */
+  public String getIdPrefix(Object instance) {
+    return null;
+  }
+
+  /**
+   * Creates a new JsogPolicy that enables JSOG for the specified types.
+   *
+   * <p>Other than that, it behaves exactly like this instace.</p>
+   *
+   * @param types Types where JSOG should be enabled
+   * @return a {@link JsogPolicy} with JSOG enabled for the specified types
+   */
+  public JsogPolicy withWhitelist(Type... types) {
+    final Set<Type> typeSet = new HashSet(Arrays.asList(types));
+    return new Wrapper(this) {
+      @Override
+      public boolean isJsogEnabled(TypeToken type) {
+        if (typeSet.contains(type.getType()) || typeSet.contains(type.getRawType())) {
+          return true;
+        }
+        return super.isJsogEnabled(type);
+      }
+    };
+  }
+
+  /**
+   * Creates a new JsogPolicy that disables JSOG for the specified types.
+   *
+   * <p>Other than that, it behaves exactly like this instace.</p>
+   *
+   * @param types Types where JSOG should be disabled
+   * @return a {@link JsogPolicy} with JSOG disabled for the specified types
+   */
+  public JsogPolicy withBlacklist(Type... types) {
+    final Set<Type> typeSet = new HashSet(Arrays.asList(types));
+    return new Wrapper(this) {
+      @Override
+      public boolean isJsogEnabled(TypeToken type) {
+        if (typeSet.contains(type.getType()) || typeSet.contains(type.getRawType())) {
+          return false;
+        }
+        return super.isJsogEnabled(type);
+      }
+    };
+  }
+
+
+  /**
+   * Creates a new JsogPolicy that prefixes IDs with the class name.
+   *
+   * <p>Examples: "Person-1", "Student-1", "Person-2", "OuterClass.Innerclass-1"</p>
+   *
+   * <p>Other than that, it behaves exactly like this instace.</p>
+   *
+   * @return a {@link JsogPolicy} which prefixes JSOG ids with the class name
+   */
+  public JsogPolicy withTypePrefix() {
+    return new Wrapper(this) {
+      @Override
+      public String getIdPrefix(Object instance) {
+        String prefix = instance.getClass().getName();
+        int indexOfLastDot = prefix.lastIndexOf(".");
+        if (indexOfLastDot != -1) {
+          prefix = prefix.substring(indexOfLastDot + 1);
+        }
+        prefix = prefix.replaceAll("\\$", ".");
+        return prefix;
+      }
+    };
+  }
+
+  /**
+   * Creates a new JsogPolicy that enabled JSOG for all data types
+   *
+   * <p>Other than that, it behaves exactly like this instace.</p>
+   *
+   * @return a {@link JsogPolicy} that enables JSOG on all data types
+   */
+  public JsogPolicy withJsogAlwaysEnabled() {
+    return new Wrapper(this) {
+      @Override
+      public boolean isJsogEnabled(TypeToken type) {
+        return true;
+      }
+    };
+  }
+
+  /**
+   * Helper class that allows easily tweaking an existing JsogPolicy by
+   * overriding the target methods
+   */
+  private class Wrapper extends JsogPolicy {
```

Comment:
```suggestion
  private static class Wrapper extends JsogPolicy {
```
Use static nested class instead of inner class for better performance

Or since this is used like `new Wrapper(this)` only, remove `wrapped` field and access enclosing instance (`JsonPolicy.this`) instead